File: 3405-count-the-number-of-special-characters-ii/count-the-number-of-special-characters-ii.java
class Solution {
    public int numberOfSpecialChars(String word) {
        int count = 0;
        Map<Character, Integer> map = new HashMap<>();
        for (int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);
            // last occurance of small letter
            if (Character.isLowerCase(ch)) {
                map.put(ch, i);
            }
            // first occurance of capital letter 
            if (Character.isUpperCase(ch) && !map.containsKey(ch)) {
                map.put(ch, i);
            }
        }
        for (int i = 0; i < 26; i++) {
            char lower = (char) ('a' + i);
            char upper = (char) ('A' + i);
            if (map.containsKey(lower) && map.containsKey(upper) && map.get(lower) < map.get(upper)) {
                count++;
            }
        }
        return count;
    }
}

File: 3405-count-the-number-of-special-characters-ii/count-the-number-of-special-characters-ii.cpp
class Solution {
public:
    int numberOfSpecialChars(string word) {
        unordered_map<char,int>umap;
        int n=word.size();
        for(int i=0;i<n;i++){
            if(isupper(word[i]) && umap.find(word[i])==umap.end()){
                umap[word[i]]=i;
            }
            if(islower(word[i])){
                umap[word[i]]=i;
            }
        }
        for(auto ele:umap){
            cout<<ele.first<<" "<<ele.second<<endl;
        }
        int count=0;
        for(char c='a';c<='z';c++){
            if(umap.find(c)!=umap.end() && umap.find(c-'a'+'A')!=umap.end()){
            if(umap[c]<umap[c-'a'+'A']){
                count++;
            }
            }
        }
        return count;
    }
};